File: python/src/dec_3/boj/사다리조작_박재환.py
n, m, h = map(int, input().split())
conn_status = [list([0]*(n+1)) for _ in range(h+1)]  # 초기 0으로 초기화 
for i in range(m):
    row, col = map(int, input().split())
    # 좌 -> 우 연결은 1로 표기
    # 우 -> 좌 연결은 2로 표기
    conn_status[row][col] = 1
    conn_status[row][col+1] = 2
min_count = 4  # 새로 놓을 수 있는 가로선의 개수는 최대 3개S

def solution():
    # 새로 놓을 수 있는 가로선의 개수는 최대 3개
    for i in range(4):
        find_new_location(i, 1, 0)
        if(min_count != 4) :
            break

    print(-1 if min_count == 4 else min_count)

def find_new_location(target_count, row_idx, new_count):
    if new_count == target_count:
        # 요구하는 사다리 조건을 만족하는지
        if is_correct():
            global min_count
            min_count = min(min_count, new_count)
        return
    
    for row in range(row_idx, h+1):
        for col in range(1, n):
            if conn_status[row][col] == 0 and conn_status[row][col+1] == 0:
                # 가로선 설치
                conn_status[row][col] = 1
                conn_status[row][col+1] = 2

                find_new_location(target_count, row, new_count + 1)

                # 가로선 제거 (백트래킹)
                conn_status[row][col] = 0
                conn_status[row][col+1] = 0

def is_correct():
    for col in range(1, n+1):
        cur_col = col
        for row in range(1, h+1):
            if conn_status[row][cur_col] == 1:
                cur_col += 1
            elif conn_status[row][cur_col] == 2:
                cur_col -= 1
        if cur_col != col:
            return False
    return True

if __name__ == "__main__":
    solution()

File: java/src/dec_3/boj/사다리조작_박재환.java
package dec_3.boj;

import java.util.*;
import java.io.*;

public class 사다리조작_박재환 {
    static BufferedReader br;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
    }
    /**
     * N개의 세로선과 M개의 가로선
     * 인접한 세로선 사이에는 가로선을 놓을 수 있다.
     * 각 세로선마다 가로선을 놓을 수 있는 위치의 개수는 H이다.
     */
    static StringTokenizer st;
    static int n,m,h;
    static int[][] connectStatus;
    static int minCnt;
    static void init() throws IOException {
        minCnt = Integer.MAX_VALUE;
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        h = Integer.parseInt(st.nextToken());
        connectStatus = new int[h+1][n +1];      // [h][col]
        for(int i=0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int rowIdx = Integer.parseInt(st.nextToken());
            int colIdx = Integer.parseInt(st.nextToken());

            // -> 방향은 1로, <- 방향은 2로 표시
            connectStatus[rowIdx][colIdx] = 1;
            connectStatus[rowIdx][colIdx+1] = 2;
        }
        solution();
    }

    static void solution() {
        for(int targetCnt = 0; targetCnt < 4 && minCnt == Integer.MAX_VALUE; targetCnt++) {
            findNewLocation(targetCnt, 1, 0);
        }

        System.out.println(minCnt == Integer.MAX_VALUE ? -1 : minCnt);
    }

    static void findNewLocation(int targetCnt, int rowIdx, int newCnt) {
        if(newCnt == targetCnt) {
            // 모두 원하는대로 방문 가능한지 검증
            if(isCorrect()) {
                minCnt = newCnt;
            }
            return;
        }

        for(int row=rowIdx; row<h+1; row++) {
            for(int col=1; col<n; col++) {  // 마지막 세로선에서는 새롭게 가로선을 추가할 수 없음
                if(connectStatus[row][col] == 0 && connectStatus[row][col+1] == 0) {
                    connectStatus[row][col] = 1;
                    connectStatus[row][col+1] = 2;
                    findNewLocation(targetCnt, row, newCnt+1);
                    connectStatus[row][col] = 0;
                    connectStatus[row][col+1] = 0;
                }
            }
        }
    }

    static boolean isCorrect() {
        for(int col=1; col<n+1; col++) {
            int curCol = col;
            for(int row=1; row<h+1; row++) {
                if(connectStatus[row][curCol] == 1) curCol++;
                else if(connectStatus[row][curCol] == 2) curCol--;
            }
            if(col != curCol) return false;
        }
        return true;
    }
}
